emily/handler/src/api/handlers/internal.rs

```diff
@@ -0,0 +1,200 @@
+        api_state = original_api_state.clone();
+
+        // Update the api status.
+        api_state.api_status = match (new_status, &original_api_state.api_status) {
+            (ApiStatus::Reorg(_), ApiStatus::Stable(_))
+            | (ApiStatus::Stable(_), ApiStatus::Reorg(_)) => new_status.clone(),
+            (ApiStatus::Stable(new_tip), ApiStatus::Stable(old_tip)) => {
+                if new_tip == old_tip {
+                    return Ok(None);
+                } else {
+                    new_status.clone()
+                }
+            }
+            // Handle trying to set the api status to reorganizing.
+            (ApiStatus::Reorg(new_reorg_tip), ApiStatus::Reorg(current_reorg_tip)) => {
+                if new_reorg_tip == current_reorg_tip {
+                    return Ok(None);
+                } else {
+                    let err_msg: String = format!("Trying to reorg with new chaintip {new_reorg_tip:?} while the api is reorganizing around the chaintip {current_reorg_tip:?}");
+                    warn!(err_msg);
+                    return Err(Error::InconsistentState(Inconsistency::ItemUpdate(err_msg)));
+                }
+            }
+        };
+
+        debug!("Changing Api state from [{original_api_state:?}] to [{api_state:?}]. Attempt {attempt_number} of maximum {MAX_SET_API_STATE_ATTEMPTS_DURING_REORG}.");
+
+        // Attempt to set the API state.
+        match accessors::set_api_state(context, &api_state).await {
+            // We successfully set the API state.
+            Ok(()) => {
+                info!("Successfully set api state: {:?}.", api_state);
+                return Ok(Some(api_state));
+            }
+            // Retry if there was a version conflict.
+            Err(Error::VersionConflict) => {
+                debug!("Failed to update API state - retrying: {api_state:?}")
+            }
+            // If some other error occured then return from here; this shouldn't
+            // happen and something has actually gone wrong.
+            e @ Err(_) => e?,
+        }
+    }
+    // Return.
+    Err(Error::InternalServer)
+}
+
+/// Handler that executes a reorg.
+///
+/// This function isn't intendeded to be exposed into any specific endpoint
+/// outside of what could maybe be a testing endpoint one day. It handles
+/// the internal requests to execute a reorg.
+pub async fn execute_reorg_handler(
+    context: &EmilyContext,
+    request: ExecuteReorgRequest,
+) -> Result<impl warp::reply::Reply, Error> {
+    info!("Executing a reorg with request {request:?}.");
+    let empty_reply = warp::reply::with_status(warp::reply(), StatusCode::NO_CONTENT);
+
+    let new_status = ApiStatus::Reorg(request.canonical_tip.clone().into());
+    match set_api_state_status(context, &new_status).await? {
+        // Do nothing if we claimed the api correctly.
+        Some(_) => {}
+        None => {
+            return Ok(empty_reply);
+        }
+    };
```

Comment:
I'm not sure if this is really better, but this can be simplified to
```rust
// [`None`] means the API state is already what we want it to be, so we
// can exit early.
if set_api_state_status(context, &new_status).await?.is_none() {
    return Ok(empty_reply);
}
```